File: test/scripts/12-many-small-files/correct.py
#!/usr/bin/env python3
"""
CORRECT: Single file append - one file, sequential I/O.
Resume: "Avoid Many Small Files"
"""
import os
import signal
import tempfile
import time
NUM_RECORDS = 2000
DURATION_SEC = int(os.environ.get("DURATION_SEC", "300"))
stop = False

def main():
    global stop
    fd, path = tempfile.mkstemp(prefix='perf-one-', suffix='.jsonl')
    print("CORRECT: Single JSONL file append")
    signal.signal(signal.SIGINT, lambda s, f: globals().__setitem__('stop', True))
    signal.signal(signal.SIGTERM, lambda s, f: globals().__setitem__('stop', True))
    start = time.perf_counter()
    laps = 0
    try:
        with os.fdopen(fd, 'w') as f:
            while not stop and (time.perf_counter() - start) < DURATION_SEC:
                f.seek(0)
                f.truncate()
                for i in range(NUM_RECORDS):
                    f.write('{"id":' + str(i) + '}\n')
                f.flush()
                f.seek(0)
                total = 0
                for line in f:
                    total += len(line)
                laps += 1
    finally:
        os.unlink(path)
    print(f"Stopped: {laps} laps")

if __name__ == "__main__":
    main()

File: test/scripts/12-many-small-files/correct.js
const fs = require('fs');
const path = require('path');
const os = require('os');
const NUM_RECORDS = 2000;
const DURATION_SEC = parseInt(process.env.DURATION_SEC || '300', 10);
const filepath = path.join(os.tmpdir(), 'perf-one-' + Date.now() + '.jsonl');
let laps = 0;
const start = Date.now();
console.log('CORRECT: Single JSONL file');
function run() {
  if (Date.now() - start >= DURATION_SEC * 1000) {
    fs.unlinkSync(filepath);
    console.log('Stopped:', laps, 'laps');
    return;
  }
  const lines = [];
  for (let i = 0; i < NUM_RECORDS; i++) lines.push('{"id":' + i + '}');
  fs.writeFileSync(filepath, lines.join('\n'));
  const data = fs.readFileSync(filepath, 'utf8');
  data.split('\n').forEach(() => {});
  laps++;
  setImmediate(run);
}
run();
process.on('SIGINT', () => { try { fs.unlinkSync(filepath); } catch(e){} process.exit(0); });
